serde/src/private/de.rs

```diff
@@ -2197,14 +2199,18 @@ impl<'a, 'de, E> MapAccess<'de> for FlatMapAccess<'a, 'de, E>
         T: DeserializeSeed<'de>,
     {
         while let Some(item) = self.iter.next() {
```

Comment:
I still find this logic hard to follow, can you at least add some comments so future contributors know what's going on